pkg/master/thirdparty_controller.go

```diff
@@ -0,0 +1,122 @@
+
+func makeThirdPartyPath(group string) string {
+	return thirdpartyprefix + group
+}
+
+// resourceInterface is the interface for the parts of the master than know how to add/remove 
+// third party resources.  Extracted into an interface for injection for testing.
+type resourceInterface interface {
+	// Remove a third party resource based on the RESTful path for that resource
+	RemoveThirdPartyResource(path string) error
+	// Install a third party resource described by 'rsrc'
+	InstallThirdPartyResource(rsrc *expapi.ThirdPartyResource) error
+	// Is a particular third party resource currently installed?
+	HasThirdPartyResource(rsrc *expapi.ThirdPartyResource) (bool, error)
+	// List all currently installed third party resources
+	ListThirdPartyResources() []string
+}
+
+// ThirdPartyController is a control loop that knows how to synchronize ThirdPartyResource objects with
+// RESTful resources which are present in the API server.
+type ThirdPartyController struct {
+	master                     resourceInterface
+	thirdPartyResourceRegistry *thirdpartyresourceetcd.REST
+}
+
+// Synchronize a single resource with RESTful resources on the master
+func (t *ThirdPartyController) SyncOneResource(rsrc *expapi.ThirdPartyResource) error {
+	hasResource, err := t.master.HasThirdPartyResource(rsrc)
+	if err != nil {
+		return err
+	}
+	if !hasResource {
+		return t.master.InstallThirdPartyResource(rsrc)
+	}
+	return nil
+}
+
+// Synchronize all resources with RESTful resources on the master
+func (t *ThirdPartyController) SyncResources() error {
+	list, err := t.thirdPartyResourceRegistry.List(api.NewDefaultContext(), labels.Everything(), fields.Everything())
+	if err != nil {
+		return err
+	}
+	return t.syncResourceList(list)
+}
+
+func (t *ThirdPartyController) syncResourceList(list runtime.Object) error {
+	existing := sets.String{}
+	switch list := list.(type) {
+	case *expapi.ThirdPartyResourceList:
+		// Loop across all schema objects for third party resources
+		for ix := range list.Items {
+			item := &list.Items[ix]
+			// extract the api group and resource kind from the schema
+			_, group, err := thirdpartyresourcedata.ExtractApiGroupAndKind(item)
+			if err != nil {
+				return err
+			}
+			// place it in the set of resources that we expect, so that we don't delete it in the delete pass
+			existing.Insert(makeThirdPartyPath(group))
+			// ensure a RESTful resource for this schema exists on the master
+			if err := t.SyncOneResource(item); err != nil {
+				return err
+			}
+		}
+	default:
+		return fmt.Errorf("expected a *ThirdPartyResourceList, got %#v", list)
+	}
+	// deletion phase, get all installed RESTful resources
+	installed := t.master.ListThirdPartyResources()
+	for _, installedAPI := range installed {
+		found := false
+		// search across the expected restful resources to see if this resource belongs to one of the expected ones
+		for _, apiPath := range existing.List() {
+			if strings.HasPrefix(installedAPI, apiPath) {
```

Comment:
Is this broken for e.g. /apis/foo.co and /apis/foo.com? Can you add to the comments examples of both of these items so that it's clear there's not a problem like this?